### PR TITLE
Improve gomod check

### PR DESCRIPTION
**Pull request by metlos**
This adds a shared action to perform the check for replacements in go.mod.
It uses `go list` command to list all the modules and then grep through its output so that we don't have to "parse" go.mod using regex on our own. This should improve the reliability of the check.

Example runs on my test repository:

https://github.com/metlos/gh-action-test/actions/runs/8344182108
https://github.com/metlos/gh-action-test/actions/runs/8344314437

### CODE REVIEW

File: gomod-check/action.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
+name: "go.mod check"
+description: "Checks that the go.mod doesn't contain replacements for 'protected' modules. Assumes go, jq and grep are available."
```
Comment (ranakan19) [inline gomod-check/action.yml:2]: curious, do you know what happens when they are not available?

File: gomod-check/action.yml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  steps:
+  - name: Checkout code from PR event
+    shell: bash
+    if: ${{ hashFiles('go.mod') != '' }}
+    env:
+      PROTECTED_MODULES: ${{ inputs.protected-modules }}
+    run: |
+      set -e
+      REGEX="("
+      for m in $(IFS=,; echo $PROTECTED_MODULES); do
```
Comment (ranakan19) [inline gomod-check/action.yml:19]: nice!

**Review by ranakan19 — APPROVED** [review]
Looks Good!

File: gomod-check/action.yml (hunk 1 above)
Comment (metlos) [reply gomod-check/action.yml:2]: The action passes when there are no protected modules: https://github.com/metlos/gh-action-test/actions/runs/8344584416

Comment (metlos) [reply gomod-check/action.yml:2]: and it also passes (or actually doesn't even run) if there is no `go.mod` in the repo: https://github.com/metlos/gh-action-test/actions/runs/8344627115

**Review by xcoulon — APPROVED** [review]

File: gomod-check/action.yml (hunk 1 above)
Comment (MatousJobanek) [inline gomod-check/action.yml:2]: Out of curiosity - what do we need the `jq` for? I don't see it to be used down in the script...

File: gomod-check/action.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        REGEX="${REGEX}${m}|"
+      done
+      REGEX="${REGEX%?})"
+
+      if go list -m all | grep --color=never -E "${REGEX}.*\s*=>"; then
```
Comment (MatousJobanek) [inline gomod-check/action.yml:24]: having `.*` and also `\s*` doesn't make much sense, does it? The first one will take any chars (including empty strings)
Let's either keep only `.*`  or use `\w+\s+`

Comment (MatousJobanek) [reply gomod-check/action.yml:24]: ok, I'm just looking at the actual content of the `go list -m all` and the last proposed regex wouldn't work :man_facepalming: 
so keeping only `.*` should be sufficient:
```
"${REGEX}.*=>";
```

**Pull request merged**